server/schedule/merge_checker.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+	log "github.com/sirupsen/logrus"
+)
+
+// MergeChecker ensures region to merge with adjacent region when size is small
+type MergeChecker struct {
+	cluster    Cluster
+	classifier namespace.Classifier
+}
+
+// NewMergeChecker creates a merge checker.
+func NewMergeChecker(cluster Cluster, classifier namespace.Classifier) *MergeChecker {
+	return &MergeChecker{
+		cluster:    cluster,
+		classifier: classifier,
+	}
+}
+
+// Check verifies a region's replicas, creating an Operator if need.
+func (m *MergeChecker) Check(region *core.RegionInfo) (*Operator, *Operator) {
+	// region size is not small enough
+	if region.ApproximateSize >= int64(m.cluster.GetMaxMergeRegionSize()) {
+		return nil, nil
+	}
+
+	// skip hot region
+	if m.cluster.IsRegionHot(region.GetId()) {
+		return nil, nil
+	}
+
+	var target *core.RegionInfo
+	prev, next := m.cluster.GetAdjacentRegions(region)
+
+	peerCount := len(region.Region.GetPeers())
+	var direction metapb.MergeDirection
+	// if is not hot region and under same namesapce
+	if prev != nil && !m.cluster.IsRegionHot(prev.GetId()) && m.classifier.AllowMerge(region, prev) {
+		// peer count should equal
+		if peerCount == len(prev.Region.GetPeers()) {
+			target = prev
+			direction = metapb.MergeDirection_Backward
+		}
+	}
+	if next != nil && !m.cluster.IsRegionHot(next.GetId()) && m.classifier.AllowMerge(region, next) {
+		// if both region is not hot, prefer the one with smaller size
+		if target == nil || target.ApproximateSize > next.ApproximateSize {
+			// peer count should equal
+			if peerCount == len(next.Region.GetPeers()) {
+				target = next
+				direction = metapb.MergeDirection_Forward
+			}
+		}
+	}
+
+	if target == nil {
+		return nil, nil
+	}
+
+	steps, err := m.matchPeers(region, target)
+	if err != nil {
+		return nil, nil
+	}
+
+	log.Debugf("try to merge region {%v} into region {%v}", region, target)
+	op1, op2 := CreateMergeRegionOperator("merge-region", region, target, direction, OpMerge, steps)
```

Comment:
I think we need to add a limit for merge.

Reply:
Make the type 'OpMerge | OpRegion | OpLeader' if the operator contains move peer or transfer leader steps.